DSpace/dspace-replicate
# Packaging Of ResourcePolicy Objects For BagIt

Resolves: https://jira.lyrasis.org/browse/DS-4480

* Use discrete Pojos (Element) for representing metadata structure as xml rather than XmlElement
* Add Policy pojo for capturing ResourcePolicies which are on DSpaceObjects
* Write policy.xml when packaging BagIt AIPs
* When reading xml, use the ElementDeserializer to allow easier reading of different Elements

--------------------------------

## Notes

The main goal of this was to package the `ResourcePolicy` objects on each `DSpaceObject` which is being packaged. The `BagItPolicyUtil` keeps a one to one mapping from the `ResourcePolicy` to xml, mapping attributes to each `ResourcePolicy` field. The `rp-action` is mapped directly by using the values defined in `Constants` to String equivalents (see `BagItPolicyUtil#actionMapper()`). Any field which is null or empty is omitted.

An example `policy.xml` :
```

<policies>
	<policy end-date="2100-01-01" group="Administrator" action="ADMIN"/>
	<policy start-date="2000-01-01" group="Anonymous" action="READ"/>
	<policy start-date="2000-01-01" group="Anonymous" action="READ_ITEM"/>
	<policy start-date="2000-01-01" group="Anonymous" action="READ_BITSTREAM"/>
	<policy name="User add" type="TYPE_CUSTOM" description="DSpace user add policy" eperson="dspace-user@localhost.localdomain" action="Add"/>
	<policy eperson="dspace-user@localhost.localdomain" action="WRITE"/>
	<policy eperson="dspace-user@localhost.localdomain" action="REMOVE"/>
	<policy eperson="dspace-user@localhost.localdomain" action="DELETE"/>
</policies>
```

There are also changes to how xml is handled by the BagIt packers. Most of this was done in anticipation of DS-4481 because the DSpace-Roles schema would not be supported as easily with how the BagIt packers were previously handling xml. JAXB is used for reading and writing xml, with pojos created for each tag, e.g. Metadata, Value, Policy.

-----------------------------------

## Testing

1. Create a new DSpace repository with the RTS and bagit enabled
   a. In your `config/dspace.cfg`, add the following lines
    ```
    include = ${module_dir}/replicate.cfg
    include = ${module_dir}/replicate-bagit.cfg
    ```
    b. Copy all configuration files from dspace-replicate to your DSpace installation module config, e.g. `config/modules/`
    c. Check that bagit packaging is enabled in your `config/modules/replicate.cfg`
    ```
    replicate.packer.pkgtype = bagit
    replicate.packer.archfmt = tar
    ```
    d. Check that `config/modules/replicate-bagit.cfg` has set all required fields for the BTR profle
    ```
    replicate-bagit.profile = beyondtherepository
    replicate-bagit.tag.bag-info.source-organization = org.dspace
    ```
    e. Check that all rts tasks are setup in `config/modules/curate.cfg`
    ```
    plugin.named.org.dspace.curate.CurationTask = \
    org.dspace.ctask.replicate.EstimateAIPSize = estaipsize, \
    org.dspace.ctask.replicate.ReadOdometer = readodometer, \
    org.dspace.ctask.replicate.TransmitAIP = transmitaip, \
    org.dspace.ctask.replicate.TransmitSingleAIP = transmitsingleaip, \
    org.dspace.ctask.replicate.VerifyAIP = verifyaip, \
    org.dspace.ctask.replicate.FetchAIP = fetchaip, \
    org.dspace.ctask.replicate.CompareWithAIP = auditaip, \
    org.dspace.ctask.replicate.RemoveAIP = removeaip, \
    org.dspace.ctask.replicate.BagItRestoreFromAIP = restorefromaip, \
    org.dspace.ctask.replicate.BagItReplaceWithAIP = replacewithaip
    ```
2. Create a Community, Collection, and Item
3. Edit the authorizations for each of the above
    a. From the Admin Panel, `Access Control > Authorizations`
    b. If needed, a group can be created through `Access Control > Group` in order to test managed groups
4. Run the `Transfer AIP(s)` curation option for the community
5. Review the `policy.xml` written for...
    a. the Community
    b. the Collection
    c. the Item and Bitstreams
6. Remove some of the authorizations changed
7. Run the `Replace all AIPs` curation option for the community
8. Review the authorizations and verify they match what exist when the initial transfer was done
